src/XrdOss/XrdOssApi.cc

```diff
@@ -254,6 +259,77 @@ int XrdOssSys::GenRemotePath(const char *oldp, char *newp)
     return 0;
 }
 
+/******************************************************************************/
+/*                                  C l o n e                                 */
+/******************************************************************************/
+/*
+  Function: Clone contents of a file from another file.
+
+  Input:    srcFile     - clone contents of this file.
+
+  Output:   Returns XrdOssOK upon success and -errno or -osserr upon failure.
+*/
+
+int XrdOssFile::Clone(XrdOssDF& srcFile)
+{
+#if defined(FICLONE)
+  if (!canClone)
+    return -EPERM;
+
+  if (fd<0)
+     return -EBADF;
+  if (srcFile.getFD() < 0)
+     return -EBADF;
+
+  if (ioctl(fd, FICLONE, srcFile.getFD())==-1)
+    return -errno;
+
+  return XrdOssOK;
+#else
+  return -ENOTSUP;
+#endif
+}
+
+int XrdOssFile::Clone(XrdOucCloneSeg cVec[], int n)
+{
+#if defined(FICLONERANGE)
+  if (!canClone)
+    return -EPERM;
+
+  if (n < 0)
+     return -EINVAL;
+
+  if (fd<0)
+     return -EBADF;
+
+  if (n == 0)
+    return XrdOssOK;
+
+   int i, j=0;
+  do{
+   int sFd = cVec[j].srcFD;
+
+   for (i = j+1; i < n && cVec[i].srcFD == sFd; i++);
+
+   for(int k=j;k<i;k++) 
+      {struct file_clone_range fr;
+       fr.src_fd = cVec[k].srcFD;
+       fr.src_offset = cVec[k].srcOffs;
+       fr.src_length = cVec[k].srcLen;
+       fr.dest_offset = cVec[k].dstOffs;
+       int rc = ioctl(fd, FICLONERANGE, &fr);
+       if (rc < 0) return -errno;
```

Comment:
If you are worried that you can't issue a reasonable message upon the error, use the new getErrMsg() API associated with the target file to relay context. Now that I said that, I see that was only fully implemented for XrdPss but with file cloning we need to implement it for XrdOss so that error text can be returned as simple errno's don't cut it any more. Further the commit that uses that API sits abandoned in the devel branch and really needs to move to master, sigh. This needs to be straightened out but it should stop the simplificaton.